File: software/python/tart/imaging/ephemeris.py
# Copyright (C) Tim Molteno 2008-2013. All rights reserved

import math
from tart.simulation.util import Util

class Ephemeris:
  GM = 3.986005e14  # earth's universal gravitational parameter m^3/s^2
  WGS84_EARTH_ROTATION_RATE = 7.2921151467e-5;  # earth rotation rate, rad/s

  def __init__(self,in_hash):
    self.a0, self.a1, self.a2, self.a3 = in_hash['a'] # alpha parameters for ionospheric model
    self.b0, self.b1, self.b2, self.b3 = in_hash['b'] # beta parameters for ionospheric model

    self.svprn = in_hash['svprn']
    self.af0 = in_hash['af0']
    self.af1 = in_hash['af1']
    self.af2 = in_hash['af2']

    self.toc = in_hash['toc'] # GpsTime

    self.iode = in_hash['iode']
    self.crs = in_hash['crs']
    self.deltan = in_hash['deltan']
    self.m0 = in_hash['m0']

    self.cuc = in_hash['cuc']
    self.ecc = in_hash['ecc']
    self.cus = in_hash['cus']
    self.roota = in_hash['roota']

    self.toe = in_hash['toe']
    self.cic = in_hash['cic']
    self.omega_c = in_hash['omega_c']
    self.cis = in_hash['cis']

    self.i0 = in_hash['i0']
    self.crc = in_hash['crc']
    self.omega = in_hash['omega']
    self.omegadot = in_hash['omegadot']

    self.idot = in_hash['idot']
    self.codes = in_hash['codes']
    self.weekno = in_hash['weekno']
    self.l2flag = in_hash['l2flag']

    self.svaccur = in_hash['svaccur']
    self.svhealth = in_hash['svhealth']
    self.tgd = in_hash['tgd']

    self.tom = in_hash['tom']
    self.fit = in_hash['fit']
    self


  def from_hash(self, in_hash):
    ret = Ephemeris()
    ret.from_hash(in_hash)
    return ret


  def to_s(self):
    return "Ephemeris SV=#{self.svprn}, toc=#{self.toc}, a0=#{self.a0}, ecc=#{self.ecc}, m0=#{self.m0}, roota=#{self.roota}"


  def to_hash(self):
    ret = {}
    ret['a']=[self.a0, self.a1, self.a2, self.a3] # alpha parameters for ionospheric model
    ret['b']=[self.b0, self.b1, self.b2, self.b3] # beta parameters for ionospheric model

    ret['svprn']=self.svprn
    ret['af0']=self.af0
    ret['af1']=self.af1
    ret['af2']=self.af2

    ret['toc']=self.toc # GpsTime

    ret['iode']=self.iode
    ret['crs']=self.crs
    ret['deltan']=self.deltan
    ret['m0']=self.m0

    ret['cuc']=self.cuc
    ret['ecc']=self.ecc
    ret['cus']=self.cus
    ret['roota']=self.roota

    ret['toe']=self.toe
    ret['cic']=self.cic
    ret['omega_c']=self.omega_c
    ret['cis']=self.cis

    ret['i0']=self.i0
    ret['crc']=self.crc
    ret['omega']=self.omega
    ret['omegadot']=self.omegadot

    ret['idot']=self.idot
    ret['codes']=self.codes
    ret['weekno']=self.weekno
    ret['l2flag']=self.l2flag

    ret['svaccur']=self.svaccur
    ret['svhealth']=self.svhealth
    ret['tgd']=self.tgd

    ret['tom']=self.tom
    ret['fit']=self.fit
    return ret


  #\brief Find the clock correction from UTC for this SV
  #     This is done ignoring leap seconds. In other words, this is modulo 1 second.
  #
  def clock_correct(self, sow):
    # more accurate calculation
    e = self.getE(self.Tsv)
    dt = Util.check_t(self.Tsv - self.toc.sow())
    dtr = -4.442807e-10 * self.ecc * self.roota * math.sin(e)   #relatavistic correction
    dtsv = self.af0 + self.af1*(dt) + self.af2*dt*dt + dtr - self.tgd
    return dtsv


  def get_tk(self, sow):
    tk = Util.check_t(sow-self.toe) #    Time from ephemeris reference epoch (1)
    if (tk < -302400.0):
      raise "Invalid time %f" % tk
    if (tk > 302400.0):
      raise "Invalid time %f" % tk
    return tk

  # http://home-2.worldonline.nl/~samsvl/satpos.htm
  # Get the eccentricity
  def getE0(self, sow):

    x=self.m0       # kepler's equation for eccentric anomaly ek
    y=self.m0 - (x-self.ecc*math.sin(x))
    x1=x
    x=y
    for i in range(0,15):
      x2=x1
      x1=x
      y1=y
      y = self.m0 - (x- self.ecc*math.sin(x))
      if (abs(y-y1)<1.0e-15):
        break
      x=(x2*y-x*y1)/(y-y1)
    ek=x    # end of det. of ecc. anomaly

    tk = self.get_tk(sow)
    n0 = math.sqrt(Ephemeris.GM/(self.roota**6)) #    Computed mean motion
    n = n0+self.deltan            #    Corrected mean motion
    mk=self.m0+n*tk  #  mean anomaly

    x=mk  #  kepler's equation for eccentric anomaly ek
    y=mk-(x- self.ecc* math.sin(x))
    x1=x
    x=y
    for i in range(0,15):
      x2=x1
      x1=x
      y1=y
      y=mk-(x- self.ecc* math.sin(x))
      if(abs(y-y1)<1.0E-15):
        break
      x=(x2*y-x*y1)/(y-y1)
    ek=x

    print "  E0 -> ", ek
    return ek

  def getE(self, sow):
    a = self.roota*self.roota
    tk = self.get_tk(sow)
    n0 = math.sqrt(Ephemeris.GM/(a**3)) #    Computed mean motion
    n = n0+self.deltan            #    Corrected mean motion
    m = self.m0+n*tk              #    Mean anomaly

    #test = self.getE0(sow)
    m = Util.rem2pi(m + Util.PI2)
    e = m
    for i in range(0,15):
      e_old = e
      e = m + self.ecc*math.sin(e)
      dE = Util.rem2pi(e - e_old)
      if (abs(dE) < 1.0e-15):
        break

    e = Util.rem2pi(e + Util.PI2)
    return e


  def get_sv_position(self, gt):
    return self.get_location(gt.sow())

  def get_sv_position_utc(self, utc_datetime):
    gpst = gps_time.GpsTime.from_time(utc_datetime)
    return self.get_location(gpst.sow())

  def get_location(self, sow):
    a = self.roota*self.roota    # Semi major axis
    tk = self.get_tk(sow)

    e = self.getE(sow)

    v = math.atan2(math.sqrt(1.0-(self.ecc**2))*math.sin(e), math.cos(e)-self.ecc)
    phi = v+self.omega
    phi = Util.rem2pi(phi)
    phi2 = 2.0*phi

    cosphi2 = math.cos(phi2)
    sinphi2 = math.sin(phi2)

    u = phi + self.cuc*cosphi2+self.cus*sinphi2
    r = a*(1.0-self.ecc*math.cos(e)) + self.crc*cosphi2+self.crs*sinphi2
    i = self.i0+self.idot*tk + self.cic*cosphi2+self.cis*sinphi2
    om = self.omega_c + (self.omegadot - Ephemeris.WGS84_EARTH_ROTATION_RATE)*tk - Ephemeris.WGS84_EARTH_ROTATION_RATE*self.toe
    om = Util.rem2pi(om + Util.PI2)
    x1 = math.cos(u)*r
    y1 = math.sin(u)*r

    x = x1*math.cos(om) - y1*math.cos(i)*math.sin(om)
    y = x1*math.sin(om) + y1*math.cos(i)*math.cos(om)
    z = y1*math.sin(i)
    return [x,y,z]

  def get_velocity(self, sow):
    loc1 = np.array(self.get_location(sow - 0.5))
    loc2 = np.array(self.get_location(sow + 0.5))
    return (loc1 - loc2)

File: software/python/tart/imaging/cpp/ephemeris.cpp
/* Copyright (C) Tim Molteno 2008-2010. All rights reserved */

#include "ephemeris.h"

#include <cmath>
#include "util.h"

using namespace std;

double ephemeris::getE(const second_of_week& sow)
{
	double t = sow._seconds;

	static double GM = 3.986005e14;  // earth's universal gravitational parameter m^3/s^2
	
	// Procedure for coordinate calculation
	double A = roota*roota;
	double tk = util::check_t(t-toe);
	double n0 = sqrt(GM/pow(A,3));
	double n = n0+deltan;
	double M = M0+n*tk;

	M = util::rem2pi(M + util::pi2);
	double E = M;
	for (int i = 0; i< 10; i++)
	{
		double E_old = E;
		E = M + ecc*sin(E);
		double dE = util::rem2pi(E - E_old);
		if (abs(dE) < 1.e-12)
			break;
	}
	E = util::rem2pi(E + util::pi2);
	return E;
}

vector3 ephemeris::get_location(const second_of_week& sow)
{		
	double t = sow._seconds;
	double A = roota*roota;
 	double tk = util::check_t(t-toe);

	double E = getE(sow);

	static double Omegae_dot = 7.2921151467e-5;  // earth rotation rate, rad/s

	double v = atan2(sqrt(1-pow(ecc,2))*sin(E), cos(E)-ecc);
	double phi = v+omega;
	phi = util::rem2pi(phi);
	double phi2 = 2.0*phi;
	
	double cosphi2 = cos(phi2);
	double sinphi2 = sin(phi2);
	
	double u = phi              + cuc*cosphi2+cus*sinphi2;
	double r = A*(1-ecc*cos(E)) + crc*cosphi2+crs*sinphi2;
	double i = i0+idot*tk       + cic*cosphi2+cis*sinphi2;
	double Omega = Omega0 + (Omegadot - Omegae_dot)*tk - Omegae_dot*toe;
	Omega = util::rem2pi(Omega + util::pi2);
	double x1 = cos(u)*r;
	double y1 = sin(u)*r;

	vector3 satp;
	satp.x = x1*cos(Omega) - y1*cos(i)*sin(Omega);
	satp.y = x1*sin(Omega) + y1*cos(i)*cos(Omega);
	satp.z = y1*sin(i);

	return satp;
}


vector3 ephemeris::get_velocity(const second_of_week& sow)
{
	vector3 ret;
#if 0
	// more accurate calculation TODO tom's velocity formula here.
	ret = foo.
	cout << "Accurate " << ret_accurate;
#else
	vector3 loc1 = get_location(sow - 0.5);
	vector3 loc2 = get_location(sow + 0.5);

	ret = (loc1 - loc2);
//	cout << "Simple " << ret;
#endif
	
	return ret;
}


#include <iostream>
using namespace std;

/*!\brief Find the clock correction from UTC for this SV
	This is done ignoring leap seconds. In other words, this is modulo 1 second.

	Tutc = Tsv - af0  -  af1  *(Tsv-Toc) - ... -  A0  - ... - leap_sec
*/
double ephemeris::clock_correct(const second_of_week& Tsv)
{
#if 1
	// more accurate calculation
	double E = getE(Tsv);
	double dt = util::check_t(Tsv - toc.sec_of_week);
	double dtr = -4.442807e-10 * ecc * roota * sin(E);	// relatavistic correction
	double dtsv = af0 + af1*(dt) + af2*dt*dt + dtr - tgd;
#else
	double dt = util::check_t(Tsv._seconds - toe);
	double dtsv = af0 + af1*(dt);
#endif
	return dtsv;
}
